### DISCUSSION TITLE
windows deployment interrupt on boot image via Cobbler 3.3.7

### DISCUSSION THREAD

**Discussion by badacook** [cobbler/cobbler#3867] [Q&A] [answered]
Hi cobbler guys

&#8195;  **I've long been looking forward to use cobbler as a unified OS deployment solution**,and try many times on cobbler 3.X.X but experience many failures.To test new versions and improvements, the newer Fedora Server Edition is generally used on vmware VM.This time, **after completing the successful automated deployment of Fedora Server 37 test,  according to the latest [wingen.rst](https://github.com/cobbler/cobbler/blob/main/docs/user-guide/wingen.rst) I tried win test windows deployment as UEFI iPXE install way**. And i had test windows10 、windows11 without any error on cobbler server ,but both interrupt on boot image ,here is the detail of my environment and processing log. From log i saw about WIMBoot compatible  of Available windows Images is no ,does this means several windows 10&&11 edition are not supported ,**which windows edition can be supported** , or **what other mistakes  I've done** .I am look forward to your correct guidance, many thanks.
[2024-12-18_135155_Sync_d2b70fd38f1d4ce4af2c1ab0d6ea67f5.log](https://github.com/user-attachments/files/18177615/2024-12-18_135155_Sync_d2b70fd38f1d4ce4af2c1ab0d6ea67f5.log)
[cobbler.log](https://github.com/user-attachments/files/18177640/cobbler.log)

```script
[root@test01 ~]# cat /etc/os-release | grep PRETTY_NAME
PRETTY_NAME="Fedora Linux 40 (Server Edition)"
[root@test01 ~]# cobbler --version
Cobbler 3.3.7
  source: ?, ?
  build time: Sun Nov 17 00:00:00 2024
[root@test01 ~]# cobbler status
ip             |target              |start            |state
192.168.80.201 |profile:Fedora_Server-37-x86_64|Tue Dec 17 19:45:21 2024|unknown/stalled
[root@test01 ~]# cobbler setting report | grep windows
windows_enabled                         : True
windows_template_dir                    : /etc/cobbler/windows
[root@test01 ~]# cat /etc/samba/smb.conf
[DISTRO]
        path = /var/www/cobbler
        guest ok = yes
        browseable = yes
        public = yes
        writeable = no
        printable = no
[root@test01 ~]# cobbler profile report --name=win11-x86_64 | grep Metadata
Automatic Installation Metadata : {'kernel': 'http://@@http_server@@/images/win11-x86_64/wimboot', 'bootmgr': 'bootmgr.exe', 'bcd': 'bcd', 'winpe': 'winpe.wim', 'answerfile': 'autounattended.xml', 'post_install_script': 'post_install.cmd'}
[root@test01 ~]# vim /var/lib/tftpboot/ipxe/default.ipxe
[root@test01 ~]# cat /var/lib/tftpboot/ipxe/default.ipxe
:win11-x86_64
kernel http://192.168.80.128/images/win11-x86_64/wimboot
initrd --name boot.sdi  http://192.168.80.128/cobbler/images/win11-x86_64/boot.sdi boot.sdi
initrd --name bootmgr.exe  http://192.168.80.128/cobbler/images/win11-x86_64/bootmgr.exe bootmgr.exe
initrd --name bcd  http://192.168.80.128/cobbler/images/win11-x86_64/bcd bcd
initrd --name winpe.wim  http://192.168.80.128/cobbler/images/win11-x86_64/winpe.wim winpe.wim
[root@test01 ~]# ls /var/www/cobbler/distro_mirror/win11/boot
bcd  bootfix.bin  bootmgr.efi  bootmgr.exe  boot.sdi  bootsect.exe  etfsboot.com  fonts  memtest.exe  pxeboot.n12  resources  winpe.wim  zh-CN
[root@test01 ~]#
```
![detail of my environment](https://github.com/user-attachments/assets/aac416eb-f56b-4903-a23e-5c95db88baee)
![Windows deployment error](https://github.com/user-attachments/assets/807d6a9a-a2df-4e48-aec0-d236a32a57c8)
![DHCP config](https://github.com/user-attachments/assets/e6376c32-7609-4550-86c8-9b5ee3d2a182)

**Comment by SchoolGuy** [comment]
@tpw56j May I again ask you to help out here? Windows is really not my strong suit.

**Reply by tpw56j** [reply]
@SchoolGuy Okay, I'll try to help.

**Comment by tpw56j** [comment]
@badacook "exec format error" means that there was an attempt to load a file not in EFI format.
Please share the tftp log:
```
journalctl -b -u tftp
```

**Reply by badacook** [reply]
@tpw56j  Hi  tpw56j 
     
![tftp log](https://github.com/user-attachments/assets/be0dab41-79e2-471a-9576-1ca65b121b26)

**Reply by tpw56j** [reply]
@badacook 
The problem may be that `ipxe.efi` requires another dhcp request to configure the network.

Try this:
1. add the following lines to `/etc/cobbler/boot_loader_conf/ipxe_menu.template`
```diff
@@ -6,6 +6,11 @@
 # Set Menu Default to Exit after timeout
 isset \${menu-default} || set menu-default $pxe_timeout_profile
 
+iseq ${platform} efi || goto not_efi &&
+:retry_dhcp
+dhcp || goto retry_dhcp
+:not_efi
+
 :Cobbler
 menu Cobbler | https://cobbler.github.io
 item local (local)
```
2. cobbler sync

**Comment by badacook** [comment]
@tpw56j Hi tpw56j
&#8195;  I had try as you suggested, but it still not work with "exec format error" . I've never felt so frustrated to learn something and try it so many times. Even  doubts about myself. Is it possible to record a video tutorial of Windows UEFI based deployment on Fedora Linux 40 (Server Edition) cobbler  server for reference.  So that learners can follow your best practice .

**Reply by tpw56j** [reply]
The correct `/var/lib/tftpboot/ipxe/default.ipxe` should not look like this:
```
[root@test01 ~]# cat /var/lib/tftpboot/ipxe/default.ipxe
:win11-x86_64
kernel http://192.168.80.128/images/win11-x86_64/wimboot
..
```
When I removed the header in this file, I got a similar error.

Also check that the `wimboot` file is accessible and has the correct size:
```
curl -sI http://192.168.80.128/images/win11-x86_64/wimboot | grep -i Content-Length
```
Content-Length: **62440**

**Reply by badacook** [reply]
![wimboot](https://github.com/user-attachments/assets/167f0dde-425d-4b62-b9c6-bb882f93e2ae)

**Reply by tpw56j** [reply]
@badacook  Please share the current `/var/lib/tftpboot/ipxe/default.ipxe`
```
# cat /var/lib/tftpboot/ipxe/default.ipxe
```

**Reply by badacook** [reply]
@tpw56j 
![default ipxe](https://github.com/user-attachments/assets/f8c4cc49-86db-4d4b-baa2-a416db76af76)

**Reply by badacook** [reply]
@tpw56j  Hi tpw56j
    I just had redeploy the same cobbler environment only change the server IP to 145 with associated configuration, and tested Fedora Server 37 automated deployment successfully, then begin to deploy windows 11 test.
![Fedora Server 37 automated deployment ](https://github.com/user-attachments/assets/47a207f2-738f-4c5d-9cff-40dd42f31ba3)

**Reply by tpw56j** [reply]
@badacook I already wrote above that such a manually modified `/var/lib/tftpboot/ipxe/default.ipxe` is the cause of the error.
Therefore, the recommendation remains the same:
1. add the following lines to `/etc/cobbler/boot_loader_conf/ipxe_menu.template`
```diff
@@ -6,6 +6,11 @@
 # Set Menu Default to Exit after timeout
 isset \${menu-default} || set menu-default $pxe_timeout_profile
 
+iseq ${platform} efi || goto not_efi &&
+:retry_dhcp
+dhcp || goto retry_dhcp
+:not_efi
+
 :Cobbler
 menu Cobbler | https://cobbler.github.io
 item local (local)
```
2. cobbler sync

**Reply by tpw56j** [reply]
@badacook And if you don't want to use the boot menu, then the correct `ipxe` file for the profile can be obtained like this:
```
curl http://192.168.80.128/cblr/svc/op/ipxe/profile/win11-x86_64
```
```
#!ipxe
kernel http://192.168.80.128/images/win11-x86_64/wimboot  
initrd --name boot.sdi http://192.168.80.128/cobbler/images/win11-x86_64/boot.sdi boot.sdi
initrd --name bootmgr.exe http://192.168.80.128/cobbler/images/win11-x86_64/bootmgr.exe bootmgr.exe
initrd --name bcd http://192.168.80.128/cobbler/images/win11-x86_64/bcd bcd
initrd --name winpe.wim http://192.168.80.128/cobbler/images/win11-x86_64/winpe.wim winpe.wim
boot
```

To use it, you need to replace the line in dhcpd.conf
```
              filename "/ipxe/default.ipxe";
```
with the line:
```
              filename "http://192.168.80.128/cblr/svc/op/ipxe/profile/win11-x86_64";
```

In my case, the OS loading looks like this:

![Снимок экрана от 2025-02-05 19-05-33](https://github.com/user-attachments/assets/381b1459-1a47-406d-8c24-c7cea8fec526)

**Reply by badacook** [reply]
@tpw56j Hi tpw56j
&#8195; As you recommended i had config the /etc/cobbler/boot_loader_conf/ipxe_menu.template and then run cobbler sync successfully. But it promote /var/lib/tftpboot/ipxe/default.ipxe: No such file or directory when i cat this file. I just want to test deploying single os with boot menu ,and the turn to multi-operation system deployment test.
![ipxe_menu template](https://github.com/user-attachments/assets/b7ee7db2-f233-4869-9957-5af5bbc533da)
![cobbler sync ](https://github.com/user-attachments/assets/2c78bb6a-ee3b-410e-b650-ce7c632e1f70)

**Reply by tpw56j** [reply]
@badacook Sorry, I misled you. You don't need to change the file `/etc/cobbler/boot_loader_conf/ipxe_menu.template`.
Revert it to its original state and run `cobbler sync`. After running this command, the file `/var/lib/tftpboot/ipxe/default.ipxe` should look like this:
```
#!ipxe

set menu-timeout 20000
set submenu-timeout ${menu-timeout}

# Set Menu Default to Exit after timeout
isset ${menu-default} || set menu-default local

:Cobbler
menu Cobbler | https://cobbler.github.io
item local (local)
item win11-x86_64 win11-x86_64
choose --default ${menu-default} --timeout ${menu-timeout} target && goto ${target}


:win11-x86_64
kernel http://192.168.80.128/images/win11-x86_64/wimboot  pause
initrd --name boot.sdi http://192.168.80.128/cobbler/images/win11-x86_64/boot.sdi boot.sdi
initrd --name bootmgr.exe http://192.168.80.128/cobbler/images/win11-x86_64/bootmgr.exe bootmgr.exe
initrd --name bcd http://192.168.80.128/cobbler/images/win11-x86_64/bcd bcd
initrd --name winpe.wim http://192.168.80.128/cobbler/images/win11-x86_64/winpe.wim winpe.wim
boot


:local
iseq ${smbios/manufacturer} HP && exit ||
sanboot --no-describe --drive 0x80

```

**Reply by badacook** [reply]
@tpw56j Hi tpw56j
&#8195; Follow your instructions to config /var/lib/tftpboot/ipxe/default.ipxe , i have The achieved the following results:
1. UEFI Secure Boot : it turn to the bios setup ,but config with UEFI Secure Boot on the cobbler vm server
2. UEFI without Secure Boot : it appears to win11 boot menu , but need to Press key to continue booting,
3. in addition：
   - it sometimes error on starcmd with "can not mount network driver" but the smbstatus is normal,even if i restart smb and cobblerd services,until i restart the Whole server
   - autounattended.xml may not be suitable, i change the OSImage install by IMAGE/INDEX , anyhow it came to setup with  autounattended.xml that seems to be successful deployment

&#8195; about the deployment menu and Press key to continue booting ,this two human interaction can it be optimize ，such cancel the second "Press key to continue booting".
furthermore a standar configration for dhcp and default menu that support Linux and Windows is looking forward to .Finally, thank you very much for your help.

**Reply by tpw56j** [reply]
1. I don't really understand how this could happen. Please share a screen just before switching to BIOS setup. On my VMs with UEFI Secure Boot I have not noticed this behavior.
2. Apparently, instead of returning the original contents of `/etc/cobbler/boot_loader_conf/ipxe_menu.template` and running `cobbler sync`, you substituted my example into `/var/lib/tftpboot/ipxe/default.ipxe`. There I really use "Press key" for debugging.
3. in addition:
    - in this case, it is most likely a samba, firewall or selinux problem.
    - I also tried using `IMAGE/INDEX`, but found that it gave unpredictable results on different versions of Windows. You can change `answerfile.template` to suit your needs.

To get rid of "Press key" restore the original contents of `/etc/cobbler/boot_loader_conf/ipxe_menu.template` and run `cobbler sync`.
Cobbler supports the ability to create dhcpd.conf for booting `systems` for both Linux and Windows, and for `profiles` you can adjust `dhcp.tempalte`. I will try to prepare a patch so that the necessary changes are out of the box.

**Reply by badacook** [reply]
@tpw56j Hi tpw56j
![UEFI Secure Boot No Media](https://github.com/user-attachments/assets/865ef579-8282-4abb-954f-179fa5cd8114)

**Reply by tpw56j** [reply]
@badacook Were you able to replace the Secure Boot keys for your virtualization type for your VM?

In my instructions I described this process only for QEMU VM, and judging by the screenshot you are using some other hypervisor. I would be glad if you shared your experience of replacing keys for other virtualization types so that it could be included in the Cobbler documentation.

**Reply by badacook** [reply]
@tpw56j I got it the client VM i use vmware vm for test, the cobbler server is QEMU VM . in addition , maybe because of compatibility issue some template did not work.such as windows menu creation. after `cobbler sync` /var/lib/tftpboot/ipxe/default.ipxe has been deleted, i was manual created it every time for test.

**Reply by tpw56j** [reply, marked as answer]
@badacook `cobbler sync` may delete `/var/lib/tftpboot/ipxe/default.ipxe` if `enable_ipxe: false` (default) in `settings.yaml`.

**Reply by badacook** [reply]
@tpw56j  thanks very much for your help!

**Reply by badacook** [reply]
@tpw56j Hi tpw56j
&#8195; I have another question about dhcp server first start, i try to config dhcp.template and copy it to creat dhcpd.conf,but dhcp server start error with exception: a name for host declaration. when i manually solve it, and run `cobbler sync` with manage_dhcp then restart dhcp service, it works well with the same dhcpd.conf without exception of host name. Can you optimize dhcp for the first startup, or do you need to customize the dhcpd.conf for the first time instead of via dhcp.template.
![DHCP first startup](https://github.com/user-attachments/assets/85f394e5-eaf6-4bcd-8c38-e1718727d875)

**Reply by tpw56j** [reply]
@badacook  To manage the dhcp server configuration, you need to:
1.  set the parameters:
- `manage_dhcp: true`
- `manage_dhcp_v4: true`
2. customize the `dhcp.template`
3. run `cobbler sync` to recreate the `/etc/dhcp/dhcpd.conf` and restart the `dhcpd` service

A copy of `dhcp.template` is not suitable as `dhcpd.conf`, `dhcpd.conf` must be created by `cobbler sync` command. 
See https://cobbler.readthedocs.io/en/latest/user-guide/dhcp-management.html#isc-dhcp
